src/jquery.uls.languagefilter.js

```diff
@@ -167,76 +167,69 @@
 		},
 
 		search: function () {
-			var langCode, scriptGroup, langNum, languagesInScript,
-				languages = $.uls.data.getLanguagesByScriptGroup( this.options.languages ),
-				query = $.trim( this.$element.val() );
-
-			this.resultCount = 0;
-			for ( scriptGroup in languages ) {
-				languagesInScript = languages[ scriptGroup ];
-				languagesInScript.sort( $.uls.data.sortByAutonym );
-				for ( langNum = 0; langNum < languagesInScript.length; langNum++ ) {
-					langCode = languagesInScript[ langNum ];
-					if ( query === '' || this.filter( langCode, query ) ) {
-						if ( this.resultCount === 0 ) {
-							// Autofill the first result.
-							this.autofill( langCode );
-						}
-
-						if ( query.toLowerCase() === langCode ) {
-							this.selectedLanguage = langCode;
-						}
-
-						if ( this.render( langCode ) ) {
-							this.resultCount++;
-						}
-					}
-				}
+			var languages = Object.keys( this.options.languages ),
+				results = [],
+				query = $.trim( this.$element.val() ).toLowerCase();
+
+			if ( query === '' ) {
+				languages.map( this.render.bind( this ) );
+				this.resultHandler( query, languages );
+				return;
 			}
 
-			// Also do a search by search API
-			if ( !this.resultCount && this.options.searchAPI && query ) {
-				this.searchAPI( query );
+			// Local search results
+			results = languages.filter( function ( langCode ) {
+				return this.filter( langCode, query );
+			}.bind( this ) );
+
+			// Use the searchAPI if available, assuming that it has superior search results.
+			if ( this.options.searchAPI ) {
+				this.searchAPI( query )
+					.done( this.resultHandler.bind( this ) )
+					.fail( this.resultHandler.bind( this, query, results ) );
 			} else {
-				this.resultHandler( query );
+				this.resultHandler( query, results );
 			}
 		},
 
 		searchAPI: function ( query ) {
-			var languageFilter = this;
+			return $.get( this.options.searchAPI, { search: query } ).then( function ( result ) {
```

Comment:
If the search api fails, it results this error:

```
jquery.uls.languagefilter.js:254 Uncaught TypeError: languageName.substring is not a function
    at LanguageFilter.autofill (jquery.uls.languagefilter.js:254)
    at LanguageFilter.resultHandler (jquery.uls.languagefilter.js:228)
```

languageName comes as rejected Promise

Reply:
This is only possible if `languageName` is truthy but not a string.

I guess on line 189, the third actual parameter might come from the failed promise.